로또 미션 제출

pkalsh opened the pull request.

pkalsh committed:
chore: empty.txt 삭제

pkalsh committed:
chore: 기능 구현 목록 작성

pkalsh committed:
test: 금액에 따른 로또 발행 테스트 작성

- 1000원 단위에 맞게 로또 발행
- 돈이 남는다면 무시
- 1000원 미만이면 예외 발생

pkalsh committed:
feat: 로또 발행 기능 구현

- LottoTickets 반환
- 가격에 관한 비즈니스 로직 책임

pkalsh committed:
refactor: LottoTickets getter 네이밍 변경

pkalsh committed:
feat: 로또 발행시 생성되는 LottoTicket 객체 구현

- 한 티켓당 로또 번호가 몇 개인지에 관한 책임
- LottoNumber VO의 콜렉션을 래핑한 객체

pkalsh committed:
feat: LottoTicket에 래핑되는 값 객체 구현

- 하나의 로또 번호의 범위에 관한 책임
- 불변성 보장

pkalsh committed:
feat: LottoTicket의 일급 컬렉션 LottoTickets 구현

- LottoTicket의 컬렉션을 래핑
- 불변성 보장

pkalsh committed:
test: 생성된 LottoNumber에 대한 제약 조건 검증 테스트

- LottoTicket 내의 LottoNumber의 개수는 6개
- LottoNumber는 1이상 45이

pkalsh committed:
feat: 당첨 번호 객체 생성 로직 구현

- 당첨 번호를 LottoNumber의 리스트로 가짐
- 보너스 번호는 따로 LottoNumber로 가짐

pkalsh committed:
feat: 로또 번호 일치 여부 판단하는 객체 ResultDiscriminator 구현

- 유저 로또 번호와 당첨 번호를 비교한 결과를 알려주는 책임
- 유저 로또 번호 LottoTickets와 당첨 번호 객체 WinningNumber를 상태로 가짐
- 실제 비교하는 로직은 LottoTickets에게 위임

pkalsh committed:
refactor: 당첨 번호 제약 조건 추가

pkalsh committed:
refactor: LottoTicket로 클래스명 변겨에 따른 메소드명 변경

pkalsh committed:
feat: LottoNumber 생성시 입력 검증 로직 추가

pkalsh committed:
test: LottoNumber VO 테스트 작성

- LottoNumber가 생성하는 후보 숫자는 1부터 45 사이
- 비정상적인 인수로 LottoNumber 생성시 예외 발생

pkalsh committed:
fix: 로또 번호 생성 범위 수정

- IntStream.range()의 endExclusive는 주어진 인자를 포함하지 않음

pkalsh committed:
feat: 당첨 번호와 발급 번호의 비교 결과를 담는 MatchCount 객체 구현

pkalsh committed:
feat: 당첨 결과에 따른 상금을 매핑하는 열거형 객체 GameResult 구현

- MatchCount를 받아 적절한 GameResult를 생성하는 evaluate 메소드
- 비교 결과 판단은 MatchCount에게 위임
- 정의된 비교 결과 이외엔 예외 발생

pkalsh committed:
test: GameResult 테스트 작성

MatchCount의 모든 메서드 동반

pkalsh committed:
feat: 일치 여부 확인하는 기능 구현

- 각 발급 티켓의 결과를 원하므로 LottoTickets에게 위임 (matchNumbers: GameResults)
- 발급 티켓과 당첨 번호의 비교 결과는 WinningNumber가 판단하도록 위임 (matchNumbers: MatchCount)
- 비교 결과를 GameResult로 변환 (evaluate: GameResult)

pkalsh committed:
refactor: LottoTickets와 WinningNumber를 래핑하던 ResultDiscriminator 객체 삭제

불필요한 wrapping

pkalsh committed:
test: 범위 체이닝 메서드 수정

- isGreaterThan -> isGreaterThanOrEqualTo
- isLessThan -> isLessThanOrEqualTo

pkalsh committed:
test: 보너스 번호가 당첨 번호에 포함되는 입력이면 에외 발생하는 테스트 작성

pkalsh committed:
test: 해시코드 테스트 커버리지 커버

pkalsh committed:
feat: 출력을 위한 description과 getter 추가

pkalsh committed:
test: getDescription 테스트

pkalsh committed:
style: 공백 맞추기

pkalsh committed:
feat: 각 결과에 대해 상금 계산 기능과 주어진 GameResult의 개수 세는 기능 추가

- calculatePrize
- count

pkalsh committed:
test: GameResults 테스트 작성

pkalsh committed:
feat: 통계 출력을 위한 dto인 WinningResult 구현

pkalsh committed:
feat: 통계 출력 담당 객체 Statistics 구현

- winningResults: 각 GameResult에 대한 결과 출력을 위한 WinningResult 리스트 반환
- calculateEarningRate: 수익률 계산
- isProfit: 수익률이 이익인지 반환

pkalsh committed:
feat: 통계 출력 담당 객체 Statistics 테스트 작성

pkalsh committed:
feat: ui 구현 및 ui를 사용하는 컨트롤러 객체 LottoApplication 구현

pkalsh committed:
test: 테스트 커버리지를 위한 LottoApplication 테스트 작성

pkalsh committed:
refactor: 패키지 재구성

jjanggyu commented on README.md:
```diff
@@ -1,5 +1,34 @@
 # 🚀 로또 1단계 - 자동
 
+## 기능 구현 목록
+### UI 로직
+- [ ] 구매 금액 입력
+- [ ] 당첨 번호 입력
+- [ ] 로또 번호 출력
+- [ ] 당첨 금액 출력
+- [ ] 수익률 출력
+###비즈니스 로직
+- [X] 로또 번호는 45까지 허용
+- [X] 로또 번호는 6자리
+- [X] 보너스 번호는 1자리
+- [ ] 당첨 금액 계산
+- [X] 로또 한 장 당 가격은 1000원
+- [X] 일치 여부 확인
+- [ ] 수익률 계산
+- [X] 로또 번호 생성
+- [X] 로또 발행
+
+## 예외 목록
+### UI 로직
+### 비즈니스 로직
+- [ ] 로또 번호 6자리 이상 입력될 시에 어떤 처리?
+- [ ] 보너스 볼 1개 이상 입력될 시에 어떤 처리?
+- [X] 당첨 번호 46 이상 입력될 시에 어떤 처리?
+- [X] 구매 금액이 1000원 단위라면 어떤 처리?
+- [X] 구매 금액이 1000원 이하라면 어떤 처리?
+- [X] 보너스 번호는 당첨 번호와 다른 번호여야 한다.
```
완성된 거 체크해주세요ㅎㅎ

pkalsh replied:
체크 했어요 ㅎ

jjanggyu commented on src/main/java/lotto/domain/MatchCount.java:
```diff
@@ -0,0 +1,36 @@
+        return matchCount == 5 && isBonusMatch;
+    }
+
+    public boolean isAllMatch() {
+        return matchCount == 6;
+    }
```
isAllMatch 대신에 isSixMatch를 쓰는 건 어떨까요? 위의 다른 것들과 통일성도 주면서..

pkalsh replied:
로또라는 실체가 분명하고 요구사항이 변경되지 않을것 같은 부분이라 전부 맞았다는 네이밍을 붙혀 봤어요.

모두 일치한다는 의미로써의 네이밍이 더 유의미하다고 생각하는데 어떻게 생각하실까요?

jjanggyu commented on src/main/java/lotto/domain/Statistics.java:
```diff
@@ -0,0 +1,35 @@
+import java.util.stream.Collectors;
+
+public class Statistics {
+    private static final int IS_EQUAL_RATE = 1;
```
이익과 손해를 구분하는 기준이라는 1이라는 숫자를 IS_EQUAL_RATE 라는 이름으로 정해준 것이 너무 좋아요! 혹시 IS_EQUAL_RATE 대신에 본전을 치다, 손익분기점의 BREAK_EVEN_RATE를 사용하는 것은 어떨까요??

pkalsh replied:
좋은 네이밍이네요! 하지만 EQUAL_RATE도 의미가 통한다고 생각해서 굳이 바꾸지는 않았습니다 :smile:

jjanggyu commented on src/main/java/lotto/domain/factory/Issuer.java:
```diff
@@ -0,0 +1,31 @@
+        return new LottoTickets(lottoTickets);
+    }
+
+    private static void validate(int numberOfTickets) {
+        if (numberOfTickets < MINIMUM_NUMBER_OF_TICKETS) {
+            throw new IllegalArgumentException(MONEY_AMOUNT_EXCEPTION_MESSAGE);
+        }
+    }
```
money를 입력 받고, TICKET_PRICE로 나누어서 티켓 수를 구하고, 티켓 수가 1보다 작을 때 "금액을 1000원 이상으로 입력해주세요"하는 것보다 money 값 자체를 TICKET_PRICE와 비교해서 1000원 이하일 때 출력하는게 더 좋은 것 같아요. 어떻게 생각하시나요??

pkalsh replied:
반영했습니다 :+1:  PurchaseAmount로 감쌌어요

jjanggyu commented on src/main/java/lotto/domain/lotto/WinningNumber.java:
```diff
@@ -0,0 +1,40 @@
+                        .count()
+        ).intValue();
+
+        boolean isBonusMatch = lottoTicket.contains(bonusNumber);
```
한 줄 띄어주면 더 좋을것 같아요!

jjanggyu commented on src/main/java/lotto/domain/result/GameResult.java:
```diff
@@ -0,0 +1,51 @@
+
+public enum GameResult {
+
+    UNDER_THREE_MATCHED(MatchCount::isUnderThree, "2개 이하 일치", 0),
+    THREE_MATCHED(MatchCount::isMatchThree, "3개 일치", 5_000),
+    FOUR_MATCHED(MatchCount::isMatchFour, "4개 일치", 50_000),
+    FIVE_MATCHED_WITHOUT_BONUS(MatchCount::isMatchFiveWithoutBonus, "5개 일치", 1_500_000),
+    FIVE_MATCHED_WITH_BONUS(MatchCount::isMatchFiveWithBonus, "5개 일치, 보너스 볼 일치", 30_000_000),
+    ALL_MATCHED(MatchCount::isAllMatch, "6개 일치", 2_000_000_000);
```
숫자 가독성을 위해서 _를 쓰는군요! 좋은 정보 배워갑니다!!

jjanggyu commented on src/main/java/lotto/domain/lotto/LottoNumber.java:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    public static List<LottoNumber> range() {
+        return IntStream.range(LOTTO_NUMBER_LOWER_BOUND, LOTTO_NUMBER_UPPER_BOUND + 1)
+                .boxed()
+                .map(LottoNumber::new)
+                .collect(Collectors.toList());
```
Stream에 이런 것도 있군요. 더 꼼꼼히 공부해야겠네요! 좋은 정보 감사합니다:)

jjanggyu reviewed (commented):
미션 수행하느라 고생 많으셨습니다!!

우선은 되게 깔끔하게 잘 짜신 것 같아요! 테스트 케이스도 자세하게 잘 나누어서 해주시고 하셔서, 리뷰 하면서 몰랐던 부분이나 제가 부족했던 부분을 많이 배워가는거 같아요!! 수고 많으셨습니다!!

pkalsh committed:
refactor: IntStream을 이용하면 int 값만큼 객체 생성하는 명령을 함수 표현식으로 가능

pkalsh committed:
fix: 콤마로만 구분하면 parseInt 할때 공백도 무시해주기 때문에 범용성 면에서 이득

pkalsh committed:
chore: 요구사항 추가

pkalsh committed:
refactor: forEach(System.out::println)으로 인덴트 감

pkalsh committed:
refactor: LottoNumber를 일급 컬렉션으로 관리하는 LottoNumbers 구현

pkalsh committed:
refactor: LottoNumbers로 리팩토링함에 따라 List로 갖고 있던 객체들의 속성 또한 리팩토링

- LottoNumbers에서 중복된 로또 번호 및 로또 번호 개수 검증

pkalsh committed:
feat: 구매 전략 기능 구현

pkalsh committed:
refactor: 로또 티켓 발행 팩토리가 이제는 구매 전략을 받아 생성

pkalsh committed:
feat: 구매 금액을 VO로 포장하여 구현

- 티켓 가격을 알고 있음
- 구매 금액과 수동 티켓 구매량을 알고 있음
- 티켓 한장 가격보다 낮은 구매 금액인지 검증
- 수동 티켓 구매량이 전체 구매량을 초과하는지 검증

pkalsh committed:
refactor: 리팩토링한 도메인에 맞게 코드 수정

pkalsh committed:
fix: 로또 번호가 똑같은 번호로 Ticket이 생성되는 버그 수정

pkalsh committed:
fix: 로또 번호 출력문 수정

pkalsh committed:
test: 테스트 수정

pkalsh committed:
test: LottoNumbers 테스트 작성

pkalsh committed:
fix: 예외 검증 세분화

pkalsh committed:
fix: 입력 문자열 파싱후 공백 제거

pkalsh committed:
test: 수동 입력 테스트 커버리지

pkalsh commented:
제가 연락이 너무 늦어 희찬님께서 개인 미션 진행인줄 알고 서로 다른 코드를 작성했네요.
세션 이후에 둘이 코드 맞춰가겠습니다

kouz95 commented on src/main/java/lotto/domain/lotto/LottoNumber.java:
```diff
@@ -0,0 +1,60 @@
+    }
+
+    public static List<LottoNumber> range() {
+        return IntStream.range(LOTTO_NUMBER_LOWER_BOUND, LOTTO_NUMBER_UPPER_BOUND + 1)
```
```suggestion
        return IntStream.rangeClosed(LOTTO_NUMBER_LOWER_BOUND, LOTTO_NUMBER_UPPER_BOUND)
```

rangeClosed를 사용하면 end를 inclusive로 처리합니다 🙂

kouz95 commented on src/main/java/lotto/domain/lotto/LottoNumber.java:
```diff
@@ -0,0 +1,60 @@
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+public class LottoNumber {
```
1부터 45까지의 로또 넘버에 한해 객체를 재사용해보는 건 어떨까요?

kouz95 commented on src/main/java/lotto/domain/result/GameResult.java:
```diff
@@ -0,0 +1,51 @@
+    private static final String UNSUPPORTED_MATCH_COUNT_EXCEPTION_MESSAGE = "일치하는 게임 결과가 없습니다.";
+
+    private final Predicate<MatchCount> matchPredicate;
+    private final String description;
```
description은 view의 관심사이지 않을까요?

pkalsh replied:
동의합니다. 리뷰 반영했습니다.

kouz95 reviewed (changes requested):
고생하셨습니다. 코드 깔끔하네요 👍
리뷰 확인해주세요~

kouz95 commented on src/test/java/lotto/LottoApplicationTest.java:
```diff
@@ -0,0 +1,35 @@
+        app.run();
+        appManual.run();
+
+        //then
+        assertThat(true).isTrue();
```
바람직한 테스트는 아닌것 같습니다. 😥
나중에 테스트는 통과하는데 왜 동작 안하지? 를 마주할 수도 있을것 같아요.

위 동작을 수행 했음에도 오류가 생기지 않았다. 정도를 테스트해보는건 어떨까요?

pkalsh replied:
아 그렇게도 테스트하는 방법이 있네요 :+1:

pkalsh committed:
[refactor] IntStream의 rangeClosed 메서드로 변경

range가 exclusive이기 때문에 가독성 면에서 rangeClosed 활용

Co-authored-by: Gyeongjun Kim <lastpuzly@gmail.com>

pkalsh committed:
test: 테스트 커버리지를 위한 테스트라도 의미 있는 테스트 작성

pkalsh committed:
refactor: LottoNumber는 오직 하나의 객체만을 만들어 재활용

pkalsh committed:
refactor: GameResult에 있는 view 로직 분

kouz95 reviewed (approved):
고생하셨습니다 🙂
깔끔하네요! 👍
